cloudevents/working-drafts/xml-format.md

```diff
@@ -200,6 +216,12 @@ Example _(XML preamble and namespace definitions omitted for brevity)_:
 </batch>
 ```
 
+The `batch` element MUST NOT contain any direct child text nodes with non-whitespace
+content, and it MUST NOT contain any XML attributes other than ones in the `xmlns` namespace.
+
+The `batch` element MAY contain child elements which are not in the namespace of
+`http://cloudevents.io/xmlformat/V1`. Any such elements MUST be ignored during processing.
```

Comment:
same comment here about ignoring unknown rather than ignoring all.

Reply:
See comment above ... if it's not "ignored" by an SDK,  what are we meant to do with it ...

None of the other Format specifications support this type of ad-hoc (format specific) extension model to the best of my knowledge.

Reply:
> See comment above ... if it's not "ignored" by an SDK, what are we meant to do with it ...

Raise an error, basically. This PR saying they should be ignored is a compromise - I would personally rather prohibit them entirely...

> None of the other Format specifications support this type of ad-hoc (format specific) extension model to the best of my knowledge.

Indeed - so I'm not sure why we should allow them here, really.

Reply:
Raising an error would mean that a newer producer could never talk to an older consumer and I think that would be a big issue - especially if a consumer can not be upgraded at all, or at least not very often. 

As for not mentioned in other formats, if not then I think that's an oversight. It jumped out at me in this spec because XML allows for extensions WITHIN spec defined metadata. Meaning, something like:
```
<ce:id myns:foo="bar"> 123 </ce:id>
```
and as long as "foo" isn't in the "ce" namespace this is perfectly legal per the extensibility of XML - and is actually one of it's benefits.

Reply:
> Raising an error would mean that a newer producer could never talk to an older consumer and I think that would be a big issue - especially if a consumer can not be upgraded at all, or at least not very often.

But that's always going to be the case with a change to a data model that allows more than it did before. I'd expect that to count as XML format v2.

> As for not mentioned in other formats, if not then I think that's an oversight.

It would be good to explore this a bit more. For example, should an SDK processing a JSON format event ignore any properties starting with "_" as they can't possibly be CE attributes?

Protobuf is interesting here as we could definitely add more fields to the .proto file and they'd be parsed as "unknown field" by most proto implementations, so the data could be passed along without interpretation... but I'd personally only want to see people use that in terms of *us* extending the protobuf schema rather than users adding arbitrary fields.

> as long as "foo" isn't in the "ce" namespace this is perfectly legal per the extensibility of XML - and is actually one of it's benefits.

It's valid XML, but that doesn't mean it has to be a valid CloudEvent, any more than this is:

```xml
<ce:this_isnt_a_valid_extension_attribute_name>oops</ce:this_isnt_a_valid_extension_attribute_name>
```

I suspect this is a philosophical disagreement in terms of how beneficial extensibility is vs strictness, where I tend to favour strictness with only *very* well-defined extensibility and I believe you tend to favour more unrestricted extensibility. It's probably not worth trying to persuade each other, but I'd be interested to hear more voices than just the three of us involved in this discussion so far. I'm definitely happy to go along with the majority.

Reply:
Using an underscore in a CE attribute name (core or extension) isn't the same thing as a new CE extension attribute since the spec mandates that all attributes (core and extensions) MUST adhere to our limited character set. So I view that as a different situation.

> I suspect this is a philosophical disagreement

Fur-sure! Whether it's specs or code, I get very nervous with a "we know what people will need" design pattern that doesn't allow for people to go beyond what the authors originally thought of because inevitably people will want more. And if they can't do any additions then they'll go find some other tech to use instead (or re-invent the wheel). I won't name names, but I can think a few projects that I view as "dead" due to this - they just may not know it yet :-)

While we joke about SOAP, their "mustUnderstand" attribute was brilliant!

Reply:
Been thinking about the case @JemDay mentioned on the call today (json binary + invalid ce json top-level attribute), and I'm wondering if in this case we basically own the schema for the body since it's a CE mime type, which means we can decide what's legal. And in this case since all top level json properties are considered CE extension attributes they must adhere to our naming rules. This feels similar to us owning the CE XML namespace. However, in JSON there's only one namespace and no real way to specify something outside of that namespace.... which is why it's a different situation.  Thoughts?

Reply:
I agree that it's a different situation in that XML namespaces give us the *option* of allowing more extensibility. It doesn't *require* us to do so, however - having a single format that allows mingling of data while none of the rest do feels slightly odd, but it's not infeasible by any means.